KAFKA-19080 The constraint on segment.ms is not enforced at topic level

m1a2st opened the pull request:
The main issue was that we forgot to set
`TopicConfig.SEGMENT_BYTES_CONFIG` to at least `1024 * 1024`, which
caused problems in tests with small segment sizes.

To address this, we introduced a new internal config:
`LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG`, allowing us to set smaller
segment bytes specifically for testing purposes.

We also updated the logic so that if a user configures the topic-level
segment bytes without explicitly setting the internal config, the
internal value will no longer be returned to the user.

In addition, we removed
`MetadataLogConfig#METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG` and added
three new internal configurations:
- `INTERNAL_MAX_BATCH_SIZE_IN_BYTES_CONFIG`
- `INTERNAL_MAX_FETCH_SIZE_IN_BYTES_CONFIG`
- `INTERNAL_DELETE_DELAY_MILLIS_CONFIG`

Reviewers: Jun Rao <junrao@gmail.com>, Chia-Ping Tsai
 <chia7712@gmail.com>

m1a2st committed:
fix the topic level segment.bytes error

m1a2st committed:
use spotlessApply

m1a2st committed:
update test default value

m1a2st committed:
wip

m1a2st commented:
Hello @junrao, @chia7712 
I have a question about this problem.
In [KAFKA-16368](https://github.com/apache/kafka/pull/18140/files), only the `ServerLogConfigs` value for `log.segment.bytes` was modified to 1 MB. However, the `TopicConfig` which in `LogConfig` default for `segment.bytes` remains 14 bytes. Many tests rely on this small `segment.bytes` value to generate a large number of segments.

I think there are two possible approaches to resolve this:

1. Add `atLeast(1024 * 1024)` validation in `LogConfig`, which would require fixing around 400 tests.
2. Add a validator in `LogManager#updateTopicConfig` to validate the change request at runtime.

junrao commented:
@m1a2st : Perhaps we could somehow allow the tests to set a small `segment.bytes`.

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
revert test change

m1a2st committed:
add new constructor for test

m1a2st committed:
add escape annotation

m1a2st committed:
fix some test

chia7712 commented:
Maybe we can add a internal config to allow tests to define small size?

m1a2st committed:
Revert "add new constructor for test"

This reverts commit 0f766c5d467fc0d35ca4d8949fe478440c75adb4.

m1a2st committed:
Revert "add escape annotation"

This reverts commit a0342edfe9994e7c4a6cd89334242c834b07d5bc.

m1a2st committed:
Revert "fix some test"

This reverts commit f70746576ee4062b9e73e5915060901be045d118.

chia7712 commented:
for example:

```java

   this.internalSegmentSize = getString(TopicConfig.INTERNAL_SEGMENT_BYTES_CONFIG);

   ...
   
    public long segmentSize() {
        if (internalSegmentSize != null) return Long.parseLong(internalSegmentSize);
        return segmentMs;
    }
```

m1a2st committed:
fix all test

m1a2st committed:
fix some test

m1a2st committed:
fix MetadataLog Test

m1a2st committed:
fix MetadataLog Test

m1a2st committed:
fix some test

m1a2st committed:
fix fail tests

m1a2st committed:
Merge remote-tracking branch 'origin/KAFKA-19080' into KAFKA-19080

m1a2st committed:
fix config def error

m1a2st committed:
fix fail test

m1a2st commented:
Thanks for @junrao, @chia7712 comments, This PR is already for review, PTAL.

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

# Conflicts:
#	core/src/main/scala/kafka/log/LogCleaner.scala
#	storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java

m1a2st committed:
Update LogCleaner file

m1a2st committed:
change segmentSize modifier

chia7712 commented on clients/src/main/java/org/apache/kafka/common/config/TopicConfig.java:
```diff
@@ -214,6 +214,9 @@ public class TopicConfig {
         "timestamps exceeds this specified threshold. This configuration is ignored if message.timestamp.type=LogAppendTime.";
 
+    // visible for testing 
+    public static final String INTERNAL_SEGMENT_BYTES_CONFIG = "internal.segment.bytes";
```
`TopicConfig` is a public class, so could you please move it to `LogConfig`?

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
addressed by comments

m1a2st committed:
addressed by comments

m1a2st committed:
update the test

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -186,12 +188,13 @@ public Optional<String> serverConfigName(String configName) {
             .define(ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_CONFIG, LONG, ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_DEFAULT, atLeast(1), LOW, ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_DOC)
-            .defineInternal(ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_CONFIG, LONG, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DEFAULT, atLeast(0), LOW, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DOC);
+            .defineInternal(ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_CONFIG, LONG, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DEFAULT, atLeast(0), LOW, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DOC)
+            .defineInternal(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, INT, null, null, LOW, ServerLogConfigs.LOG_SEGMENT_BYTES_DOC);
```
Could we document that this config is for testing?

chia7712 commented:
@m1a2st could you please check the failed tests?

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
remove unused import

junrao commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -583,35 +584,72 @@ object KafkaMetadataLog extends Logging {
+  }
+
+  // visible for testing
+  def internalApply(
```
Instead of having an `internalApply`, could we just use the existing `apply` and add `INTERNAL_SEGMENT_BYTES_CONFIG` to `MetadataLogConfig`?

chia7712 replied:
+1 to move the internal config to `MetadataLogConfig`, and it would be better to wait #19465 extracting the metadata-related configs from other class to `MetadataLogConfig`

junrao replied:
I just realized that metadata log uses a different approach to allow tests to use a smaller segment bytes than allowed in production. That approach defines the original segment byte config with a small minimal requirement, but adds METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG to enforce the actual minimal requirement in production. This new config could be changed in tests to allow for smaller minimal bytes. The benefit of this approach is that it allows the existing config to be used directly to set a smaller value for tests. The downside is that the doc for min value is inaccurate and the validation is done through a customized logic.

It would be useful to pick the same strategy between metadata log and regular log. The metadata log approach seems slightly better since it's less intrusive. We could fix the inaccurate min value description for production somehow.

chia7712 replied:
Excuse me, the strategy used by metadata log is to add a "internal" config (`METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG`) to change the (metadata) segment size in testing, and that is what we want to address in this PR - we add a "internal" config for regular log, and so the test can use the "smaller" segment size.

junrao replied:
I am just saying that we now have two different ways to achieve the same goal. In the metadata log approach, you set the desired value through the original config, which is segment.bytes. You then set an internal config to change the min constraint.

The approach in this PR is to set the desired value through a different internal config. 

It would be useful to choose same approach for both the metadata log and the regular log.

chia7712 replied:
Personally, I prefer the approach of adding "internal".xxx config as it provide better user experience for public configs, allowing users to see the "correct" min value. Additionally, we can remove the customized logic of validation.

In short, I suggest to add following changes to this PR.

1. remove `METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG`
2. remove `MetadataLogConfig#logSegmentMinBytes`
3. add `internal.metadata.log.segment.bytes`
4. customize `MetadataLogConfig#logSegmentBytes` as following code
```java
    public int logSegmentBytes() {
        if (internalSogSegmentBytes != null) return internalSogSegmentBytes;
        return logSegmentBytes;
    }
```

m1a2st replied:
Thanks for @chia7712, @junrao comments, addressed it :)

junrao replied:
Hmm, why do we need `internalApply()`? MetadataLogConfig has both the product and internal segment bytes configs and we could just pass both into LogConfig in `apply()`, right?

junrao reviewed (commented):
@m1a2st : Thanks for the PR. Left a couple of comments.

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

# Conflicts:
#	core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

m1a2st committed:
move other internal config

m1a2st committed:
update KafkaMetadataLog apply flow

m1a2st committed:
fix compile error

m1a2st committed:
fix fail test

m1a2st committed:
fix fail test

m1a2st committed:
revert unused change

m1a2st committed:
revert unused change

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -85,14 +82,13 @@ public class MetadataLogConfig {
             .define(METADATA_SNAPSHOT_MAX_INTERVAL_MS_CONFIG, LONG, METADATA_SNAPSHOT_MAX_INTERVAL_MS_DEFAULT, atLeast(0), HIGH, METADATA_SNAPSHOT_MAX_INTERVAL_MS_DOC)
             .define(METADATA_LOG_DIR_CONFIG, STRING, null, null, HIGH, METADATA_LOG_DIR_DOC)
             .define(METADATA_LOG_SEGMENT_BYTES_CONFIG, INT, METADATA_LOG_SEGMENT_BYTES_DEFAULT, atLeast(Records.LOG_OVERHEAD), HIGH, METADATA_LOG_SEGMENT_BYTES_DOC)
-            .defineInternal(METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG, INT, METADATA_LOG_SEGMENT_MIN_BYTES_DEFAULT, atLeast(Records.LOG_OVERHEAD), HIGH, METADATA_LOG_SEGMENT_MIN_BYTES_DOC)
```
We need to set the constraint for METADATA_LOG_SEGMENT_BYTES_CONFIG to be at least 8MB. 

Also, I thought the plan is to remove METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG, but add sth like METADATA_INTERNAL_LOG_SEGMENT_BYTES_CONFIG to match the design in LogConfig?

chia7712 replied:
@junrao I have discussed with @m1a2st offline, and he will update the PR tomorrow

chia7712 replied:
@m1a2st What do you think about the `METADATA_INTERNAL_LOG_SEGMENT_BYTES_CONFIG`? This PR currently add only `INTERNAL_LOG_SEGMENT_BYTES_CONFIG` to represent server-level log config. That means users can't configure "small" metadata log without impacting other logs.

m1a2st replied:
I think we shouldn’t add this unless it’s actually needed by the test. Given that it’s an internal configuration, it might be more appropriate to introduce it only when a concrete use case arises.

junrao commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -585,32 +585,62 @@ object KafkaMetadataLog extends Logging {
+    val metadataLog: KafkaMetadataLog = createMetadataLog(topicPartition, topicId, dataDir, time, scheduler, config, nodeId, defaultLogConfig)
+
+    // Print a warning if users have overridden the internal config
+    if (config.logSegmentBytes() != KafkaRaftClient.MAX_BATCH_SIZE_BYTES) {
```
Hmm,  we don't need this if we add the constraint directly to METADATA_LOG_SEGMENT_BYTES_CONFIG, right?

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -112,15 +108,15 @@ public class MetadataLogConfig {
      */
     public MetadataLogConfig(int logSegmentBytes,
-                             int logSegmentMinBytes,
+                             int internalLogSegmentMinBytes,
```
Hmm, this should be renamed to internalLogSegmentBytes since it's no longer the minimum, right?

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A few more comments.

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
temp

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
completed the feature

m1a2st committed:
fix fail test

m1a2st committed:
fix fail test

chia7712 commented on core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala:
```diff
@@ -601,7 +602,7 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
     assertEquals(topicIds(topic1), createResult.topicId(topic1).get())
     assertFutureThrows(classOf[TopicAuthorizationException], createResult.topicId(topic2))
-    
+
```
please revert unrelated change

chia7712 commented on core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala:
```diff
@@ -567,7 +568,7 @@ class SaslSslAdminIntegrationTest extends BaseAdminIntegrationTest with SaslSetu
 
     val topics = Seq(topic1, topic2)
-    val configsOverride = Map(TopicConfig.SEGMENT_BYTES_CONFIG -> "100000").asJava
+    val configsOverride = Map(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG -> "100000").asJava
```
this test case is used to verify the custom topic-level config, so we can increase the value to make test pass.

chia7712 commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -52,14 +51,14 @@ public class MetadataLogConfig {
     public static final String METADATA_LOG_SEGMENT_BYTES_DOC = "The maximum size of a single metadata log file.";
     public static final int METADATA_LOG_SEGMENT_BYTES_DEFAULT = 1024 * 1024 * 1024;
 
+    public static final String INTERNAL_METADATA_LOG_SEGMENT_BYTES_CONFIG = "internal.metadata.log.segment.bytes";
```
Do we need another internal config? User can configure `internal.segment.bytes` if they want to create small segment, right?

m1a2st committed:
addressed by comments

m1a2st committed:
remove metadataLogConfig

m1a2st committed:
complete

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
revert to trunk

m1a2st committed:
using the ServerLogConfigs instead of Topic level

m1a2st committed:
fix test

m1a2st committed:
fix test

m1a2st committed:
fix test

m1a2st committed:
temp test

m1a2st committed:
fix PurgeRepartitionTopicIntegrationTest test

m1a2st committed:
spotless

m1a2st committed:
spotless

m1a2st committed:
fix test

m1a2st committed:
fix some error

m1a2st committed:
fix some error

m1a2st committed:
fix some error

m1a2st committed:
fix some error

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
reset to trunk

m1a2st committed:
temp

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -112,15 +110,14 @@ public class MetadataLogConfig {
                              long deleteDelayMillis) {
         this.logSegmentBytes = logSegmentBytes;
-        this.logSegmentMinBytes = logSegmentMinBytes;
+        this.internalLogSegmentBytes = logSegmentBytes;
```
Could we add a comment that this constructor is for test only? Could we also rename logSegmentBytes to internalLogSegmentBytes?

junrao commented on core/src/test/scala/unit/kafka/server/ControllerConfigurationValidatorTest.scala:
```diff
@@ -73,15 +74,15 @@ class ControllerConfigurationValidatorTest {
     val config = new util.TreeMap[String, String]()
     config.put(SEGMENT_JITTER_MS_CONFIG, "1000")
-    config.put(SEGMENT_BYTES_CONFIG, "67108864")
+    config.put(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, "67108864")
```
This change is unnecessary, right? Ditto below.

junrao commented on core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala:
```diff
@@ -101,27 +101,27 @@ class DynamicConfigChangeTest extends KafkaServerTestHarness {
     val oldSegmentSize = 1000
     val logProps = new Properties()
-    logProps.put(TopicConfig.SEGMENT_BYTES_CONFIG, oldSegmentSize.toString)
+    logProps.put(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, oldSegmentSize.toString)
```
Is this change needed since we don't need a small segment size? Ditto below.

junrao commented on core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala:
```diff
@@ -1103,7 +1090,7 @@ object KafkaMetadataLogTest {
     )
 
-    val metadataLog = KafkaMetadataLog(
+    val metadataLog = KafkaMetadataLog.apply(
```
Is this change needed?

junrao commented on metadata/src/test/java/org/apache/kafka/controller/ReplicationControlManagerTest.java:
```diff
@@ -905,7 +904,7 @@ public void testCreateTopicsWithPolicy() {
             new CreateTopicPolicy.RequestMetadata("baz", null, null,
                 Map.of(0, List.of(2, 1, 0)),
-                Map.of(SEGMENT_BYTES_CONFIG, "12300000")),
+                Map.of("internal.segment.bytes", "12300000")),
```
Is this change needed? Ditto below.

junrao commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorServiceTest.java:
```diff
@@ -1999,7 +1999,7 @@ public void testShareStateTopicConfigs() {
             TopicConfig.CLEANUP_POLICY_CONFIG,
             TopicConfig.COMPRESSION_TYPE_CONFIG,
-            TopicConfig.SEGMENT_BYTES_CONFIG,
+            "internal.segment.bytes",
```
Hmm, this doesn't look right. service.shareGroupStateTopicConfigs() doesn't include this internal config.

junrao commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java:
```diff
@@ -114,7 +114,7 @@ public final boolean conditionMet() {
                 return config.get(TopicConfig.CLEANUP_POLICY_CONFIG).value().equals(TopicConfig.CLEANUP_POLICY_DELETE)
                         && config.get(TopicConfig.SEGMENT_MS_CONFIG).value().equals(PURGE_INTERVAL_MS.toString())
-                        && config.get(TopicConfig.SEGMENT_BYTES_CONFIG).value().equals(PURGE_SEGMENT_BYTES.toString());
+                        && config.get("internal.segment.bytes").value().equals(PURGE_SEGMENT_BYTES.toString());
```
Could we use the constant val ? Ditto below.

junrao commented on streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:
```diff
@@ -248,7 +247,7 @@ public void consumerConfigMustContainStreamPartitionAssignorConfig() {
         props.put(StreamsConfig.WINDOW_STORE_CHANGE_LOG_ADDITIONAL_RETENTION_MS_CONFIG, 7L);
         props.put(StreamsConfig.APPLICATION_SERVER_CONFIG, "dummy:host");
-        props.put(StreamsConfig.topicPrefix(TopicConfig.SEGMENT_BYTES_CONFIG), 100);
+        props.put(StreamsConfig.topicPrefix("internal.segment.bytes"), 100);
```
Could we use the constant val ? Ditto below.

m1a2st replied:
We can't add the storage dependency to the Streams module because the Streams module targets Java 11, while the Storage module requires Java 17.

chia7712 replied:
this is used to test the updated configs get returned, so we can increase the configured value instead of using internal config.

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A few more comments.

m1a2st committed:
reset to trunk

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
completed

m1a2st committed:
update some test

m1a2st commented:
Hello @chia7712, @junrao, Sorry for the late reply.

I’ve updated the patch. Developers can still use the internal config at both the server and topic levels. I also updated the logic so that if a user configures the topic-level config without specifying the internal one, we will no longer return the internal config to the user.

m1a2st committed:
addressed by comments

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -186,12 +190,13 @@ public Optional<String> serverConfigName(String configName) {
             .define(ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_CONFIG, LONG, ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_DEFAULT, atLeast(1), LOW, ServerLogConfigs.LOG_DIR_FAILURE_TIMEOUT_MS_DOC)
-            .defineInternal(ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_CONFIG, LONG, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DEFAULT, atLeast(0), LOW, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DOC);
+            .defineInternal(ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_CONFIG, LONG, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DEFAULT, atLeast(0), LOW, ServerLogConfigs.LOG_INITIAL_TASK_DELAY_MS_DOC)
+            .defineInternal(ServerLogConfigs.INTERNAL_LOG_SEGMENT_BYTES_CONFIG, INT, null, null, MEDIUM, ServerLogConfigs.INTERNAL_LOG_SEGMENT_BYTES_DOC);
```
Hmm, with this, we should be able to update internal.log.segment.bytes for brokers, right? But the following failed.

```
bin/kafka-configs.sh --bootstrap-server localhost:9092 --alter --add-config internal.log.segment.bytes=2000 --entity-type brokers --entity-default
Error while executing config command with args '--bootstrap-server localhost:9092 --alter --add-config internal.log.segment.bytes=2000 --entity-type brokers --entity-default'
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Cannot update these configs dynamically: Set(internal.log.segment.bytes)
```

chia7712 replied:
>  with this, we should be able to update internal.log.segment.bytes for brokers, right?

The configurations listed in `SERVER_CONFIG_DE`F aren't always reconfigurable at runtime. For instance, `log.dirs` is also defined there, but it definitely cannot be changed while the server is running.

Furthermore, it seems unnecessary to make all internal configurations reconfigurable for now.

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -253,7 +258,8 @@ public Optional<String> serverConfigName(String configName) {
                 .define(TopicConfig.REMOTE_LOG_COPY_DISABLE_CONFIG, BOOLEAN, false, MEDIUM, TopicConfig.REMOTE_LOG_COPY_DISABLE_DOC)
-                .define(TopicConfig.REMOTE_LOG_DELETE_ON_DISABLE_CONFIG, BOOLEAN, false, MEDIUM, TopicConfig.REMOTE_LOG_DELETE_ON_DISABLE_DOC);
+                .define(TopicConfig.REMOTE_LOG_DELETE_ON_DISABLE_CONFIG, BOOLEAN, false, MEDIUM, TopicConfig.REMOTE_LOG_DELETE_ON_DISABLE_DOC)
+                .defineInternal(INTERNAL_SEGMENT_BYTES_CONFIG, INT, null, null, MEDIUM, INTERNAL_SEGMENT_BYTES_DOC);
```
Could we avoid listing the internal config in tools like kafka-topics and kafka-config?

m1a2st replied:
Update, test on my local machine
![CleanShot 2025-05-15 at 23 32 39@2x](https://github.com/user-attachments/assets/15933d4a-0492-486c-aec3-d4b54cdb7509)

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -101,26 +96,24 @@ public class MetadataLogConfig {
                              long deleteDelayMillis) {
-        this.logSegmentBytes = logSegmentBytes;
-        this.logSegmentMinBytes = logSegmentMinBytes;
+        this.logSegmentBytes = internalLogSegmentBytes;
```
We don't need this line, right?

chia7712 replied:
Maybe we don't need `internalLogSegmentBytes` in `MetadataLogConfig`. This constructor is used by testing, so we don't need to check the value. Another constructor accepting `AbstractConfig` can be modified like below.
```java
this.logSegmentBytes = Objects.requireNonNullElseGet(config.getInt(ServerLogConfigs.INTERNAL_LOG_SEGMENT_BYTES_CONFIG), () -> config.getInt(METADATA_LOG_SEGMENT_BYTES_CONFIG));
```

and then in constructing props by `MetadataLogConfig` we can use `LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG` directly as the values in `MetadataLogConfig` are already verified.
```java
props.setProperty(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
```

m1a2st replied:
I removed the `internalLogSegmentBytes` property, however this constructor doesn't validate `logSegmentBytes` is greater than min value. thus, I'm still passing internalLogSegmentBytes as a parameter name."

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A few more comment.

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -52,10 +52,6 @@ public class MetadataLogConfig {
             "configuration. The Kafka node will generate a snapshot when either the maximum time interval is reached or the " +
             "maximum bytes limit is reached.";
 
-    public static final String METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG = "metadata.log.segment.min.bytes";
```
@mumrah : Does this change sound reasonable to you?

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/KafkaConfigSchema.java:
```diff
@@ -166,6 +166,9 @@ public Map<String, ConfigEntry> resolveEffectiveTopicConfigs(
         ConfigDef configDef = configDefs.getOrDefault(ConfigResource.Type.TOPIC, EMPTY_CONFIG_DEF);
         HashMap<String, ConfigEntry> effectiveConfigs = new HashMap<>();
         for (ConfigDef.ConfigKey configKey : configDef.configKeys().values()) {
+            if (configKey.internalConfig && !dynamicTopicConfigs.containsKey(configKey.name)) {
```
Please add comments

m1a2st committed:
addressed by comments

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
addressed by comments

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -84,15 +80,13 @@ public class MetadataLogConfig {
             .define(METADATA_LOG_DIR_CONFIG, STRING, null, null, HIGH, METADATA_LOG_DIR_DOC)
-            .define(METADATA_LOG_SEGMENT_BYTES_CONFIG, INT, METADATA_LOG_SEGMENT_BYTES_DEFAULT, atLeast(Records.LOG_OVERHEAD), HIGH, METADATA_LOG_SEGMENT_BYTES_DOC)
-            .defineInternal(METADATA_LOG_SEGMENT_MIN_BYTES_CONFIG, INT, METADATA_LOG_SEGMENT_MIN_BYTES_DEFAULT, atLeast(Records.LOG_OVERHEAD), HIGH, METADATA_LOG_SEGMENT_MIN_BYTES_DOC)
+            .define(METADATA_LOG_SEGMENT_BYTES_CONFIG, INT, METADATA_LOG_SEGMENT_BYTES_DEFAULT, atLeast(1024 * 1024), HIGH, METADATA_LOG_SEGMENT_BYTES_DOC)
```
Hmm, the min used to be 8MB, right?

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -446,6 +460,10 @@ public static List<String> configNames() {
     }
 
+    public static List<String> excludeInternalConfigNames() {
+        return CONFIG.names().stream().filter(name -> !name.contains(INTERNAL_PREFIX)).sorted().toList();
```
Could we use `configKey.internalConfig` instead of relying on the prefix name?

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -140,11 +139,17 @@ public Optional<String> serverConfigName(String configName) {
             TopicConfig.REMOTE_LOG_STORAGE_ENABLE_CONFIG,
             TopicConfig.REMOTE_LOG_DELETE_ON_DISABLE_CONFIG,
             TopicConfig.REMOTE_LOG_COPY_DISABLE_CONFIG,
+            INTERNAL_SEGMENT_BYTES_CONFIG,
```
Hmm, do we plan to support INTERNAL_SEGMENT_BYTES_CONFIG at the broker level? If yes, we need to remove this line. If not, we need to exclude INTERNAL_SEGMENT_BYTES_CONFIG from SERVER_CONFIG_DEF, right?

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A few more comments.

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
addressed by comments

m1a2st committed:
fix fail test

junrao commented on server-common/src/main/java/org/apache/kafka/server/config/ServerTopicConfigSynonyms.java:
```diff
@@ -50,6 +50,7 @@ public final class ServerTopicConfigSynonyms {
     // Topic configs with no mapping to a server config can be found in `LogConfig.CONFIGS_WITH_NO_SERVER_DEFAULTS`
     public static final Map<String, List<ConfigSynonym>> ALL_TOPIC_CONFIG_SYNONYMS = Utils.mkMap(
         sameNameWithLogPrefix(TopicConfig.SEGMENT_BYTES_CONFIG),
+        sameName("internal.log." + TopicConfig.SEGMENT_BYTES_CONFIG),
```
This is not really the same name. Could we add a method like sameNameWithInternalLogPrefix?

junrao commented on core/src/test/scala/unit/kafka/log/LogConfigTest.scala:
```diff
@@ -52,8 +52,13 @@ class LogConfigTest {
-        val serverConfigOpt = LogConfig.serverConfigName(config)
-        serverConfigOpt.isPresent && (serverConfigOpt.get != null)
+        // this internal config naming pattern is not as same as a default pattern
+        if (config.equals(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG)) {
```
Why do we need this? ALL_TOPIC_CONFIG_SYNONYMS includes INTERNAL_SEGMENT_BYTES_CONFIG.

chia7712 replied:
I guess this change is related to the comment (https://github.com/apache/kafka/pull/19371#discussion_r2091933863)

Maybe we can remove the test `ensureNoStaticInitializationOrderDependency` and `CONFIGS_WITH_NO_SERVER_DEFAULTS`. 

They were used to test the bug caused by the initialization order between `KafkaConfig` and `LogConfig` (see https://github.com/apache/kafka/commit/0ea540552a52b3265c48aacbb9790ea6e71431a5). 
Fortunately, the dependence was gone when rewriting `LogConfig` by java (see https://github.com/apache/kafka/commit/96d9710c17b34d4e80259f09845d21de66a5efaf) :)

m1a2st replied:
Thanks for information, I think we can remove this test .

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A couple of more comments.

junrao commented:
It's useful to check if metadata.log.segment.min.bytes is used in any ducktape tests.

m1a2st committed:
addressed by comment

m1a2st committed:
revert broker setting

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
using internal config

m1a2st committed:
fix testDynamicLogConfigs

m1a2st committed:
add the comment

m1a2st commented:
> It's useful to check if metadata.log.segment.min.bytes is used in any ducktape tests.

I’ve checked all the tests in the Kafka project, and none of them use this configuration.

junrao commented on server-common/src/main/java/org/apache/kafka/server/config/ServerTopicConfigSynonyms.java:
```diff
@@ -47,9 +48,11 @@ public final class ServerTopicConfigSynonyms {
         sameNameWithLogPrefix(TopicConfig.SEGMENT_BYTES_CONFIG),
+        // Due to the internal.segment.bytes is in storage module, thus we could not use the 
+        // LogConfig#INTERNAL_SEGMENT_BYTES_CONFIG directly. We need to use the string value instead.
+        sameNameWithInternalLogPrefix("internal.segment.bytes"),
```
Is the intention to support `internal.segment.bytes` at the broker level? If so, should we add it to `KafkaConfig.extractLogConfigMap` too?

Currently, if internal.log.segment.bytes is set at the broker, it won't be picked up by a new user topic, but it will be picked up by the kraft topic.

Thinking a bit more. Do we need to support internal.segment.bytes at the broker level? For testing, is it enough to just support internal.segment.bytes at the topic level?

chia7712 replied:
We currently don't have small segment tests for Kraft topics. Therefore, a more effective way to streamline this PR would be to support topic-level configuration only. This means the new internal configuration (`internal.segment.bytes`) would apply exclusively to user topics.

We can revisit this config for Kraft topics in the future if it needs.

m1a2st replied:
In the previous solution, we required the server-level `internal.log.segment.bytes` to configure `MetadataLogConfig`.
However, since we now always use `internal.segment.bytes` in the `KafkaMetadataLog.apply` method, this server-level config is no longer necessary and can be removed.

chia7712 replied:
@m1a2st 

Do you mean this PR supports only topic-level internal config? If so, we can't use small segment for Kraft topic as it takes only server-level config. Do I understand correctly?

m1a2st replied:
Yes, you are right.

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. One more comment. Also, are the test failures related?

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st committed:
addressed by comments

junrao commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -589,8 +589,9 @@ object KafkaMetadataLog extends Logging {
-    props.setProperty(TopicConfig.SEGMENT_MS_CONFIG, config.logSegmentMillis.toString)
+    // Since MetadataLogConfig validates the log segment size using a minimum allowed value,
+    // we can safely use `internal.segment.bytes` to configure it instead of relying on `segment.bytes`.
+    props.setProperty(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
```
This could be confusing. If someone turns on the debug logging to see the config for the metadata topic, it will be surprising to see INTERNAL_SEGMENT_BYTES_CONFIG being set, even though it's not set by the users.

chia7712 replied:
We don't need this change now, as the internal config for kraft topic is removed. Additionally, the min size of kraft topic segment (8MB) is larger than size of regular segment (1MB). Hence, it is pretty safe to set `props.setProperty(TopicConfig.SEGMENT_MS_CONFIG, config.logSegmentMillis.toString)`

junrao replied:
Hmm, does that work? This will fail `LogConfig.validate(props)` if a test passes in a small log segment value, right?

chia7712 replied:
> This will fail LogConfig.validate(props) if a test passes in a small log segment value, right?

This won't happen because this PR removes support for configuring small segments for Kraft topics. The minimum Kraft topic size is now 8MB, which is large enough to pass LogConfig.validate(props). 

We've removed the configuration directly to streamline the PR, as there are no existing tests in code base for small segment Kraft topics.

If configuring small segments for kraft topic becomes necessary, we will need to reintroduce broker-level internal log configuration, as topic-level configuration is not available for kraft topic.

junrao replied:
Hmm, we have tests using the following constructor to pass in a small segmentBytes value. 

```
    public MetadataLogConfig(int internalLogSegmentBytes,
                             long logSegmentMillis,
                             long retentionMaxBytes,
                             long retentionMillis,
                             int maxBatchSizeInBytes,
                             int maxFetchSizeInBytes,
                             long deleteDelayMillis) {
```

This will fail the validation in the following if it's passed in as TopicConfig.SEGMENT_BYTES_CONFIG, right? 
```
    LogConfig.validate(props)
    val defaultLogConfig = new LogConfig(props)
```

chia7712 replied:
@junrao You're right. To support smaller segments for Kraft topics, we could reintroduce the broker-level `internal.log.segment.bytes` configuration. However, this duplicates the functionality of the topic-level `internal.segment.bytes` and would affect all topics on the server. 

Perhaps we should revisit `internal.metadata.log.segment.bytes` instead. The regular topics could utilize topic-level `internal.segment.bytes` for smaller segments, while Kraft topics would use the broker-level `internal.metadata.log.segment.bytes`. This approach would require adding if-else logic when converting Kraft configurations to log configurations.
```scala
props.setProperty(if config.useInternal LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG else LogConfig.SEGMENT_BYTES_CONFIG,  config.logSegmentBytes.toString)
```


For another, we should move the test-only constructor out of the production code. This will prevent it from being used accidentally in a production environment. This could be considered as a follow-up ticket
```java
public MetadataLogConfig(int internalLogSegmentBytes,
                             long logSegmentMillis,
                             long retentionMaxBytes,
                             long retentionMillis,
                             int maxBatchSizeInBytes,
                             int maxFetchSizeInBytes,
                             long deleteDelayMillis) {
```

chia7712 replied:
> For another, we should move the test-only constructor out of the production code. This will prevent it from being used accidentally in a production environment. This could be considered as a follow-up ticket

We should probably address this in this PR. Otherwise, the `MetadataLogConfig` constructor that takes `internalLogSegmentBytes` will seem a bit strange in the production code.

see https://github.com/apache/kafka/blob/trunk/group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorConfigTest.java#L285 as example

m1a2st replied:
I believe we shouldn’t move this constructor outside of `MetadataLogConfig`, because the public `MetadataLogConfig(AbstractConfig config)` constructor doesn’t allow us to set properties like `maxBatchSizeInBytes`, `maxFetchSizeInBytes`, or `deleteDelayMillis`.

chia7712 replied:
> maxBatchSizeInBytes, maxFetchSizeInBytes, or deleteDelayMillis.

Maybe it is time to introduce internal configs for them? That make `MetadataLogConfig` consistent with other config classes which allow to configure all variables.

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -446,6 +447,14 @@ public static List<String> configNames() {
         return CONFIG.names().stream().sorted().toList();
     }
 
+    public static List<String> excludeInternalConfigNames() {
```
excludeInternalConfigNames => nonInternalConfigNames ?

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A couple of more comments.

chia7712 commented:
@m1a2st please fix the conflicts

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

# Conflicts:
#	core/src/test/scala/integration/kafka/api/SaslSslAdminIntegrationTest.scala

m1a2st committed:
addressed by comments

m1a2st committed:
fix fail test

m1a2st committed:
fix fail test `testFromPropsInvalid`

junrao commented on core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala:
```diff
@@ -1166,4 +1161,25 @@ object KafkaMetadataLogTest {
+    logSegmentMillis: Long, 
+    retentionMaxBytes: Long, 
+    retentionMillis: Long, 
+    maxBatchSizeInBytes: Int, 
```
Should these three be named internal? Ditto in other test.

junrao commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -589,8 +589,10 @@ object KafkaMetadataLog extends Logging {
-    props.setProperty(TopicConfig.SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
-    props.setProperty(TopicConfig.SEGMENT_MS_CONFIG, config.logSegmentMillis.toString)
+    if (config.internalSegmentBytes() != null)
+      props.setProperty(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
```
Hmm, `config.logSegmentBytes` should be `config.internalLogSegmentBytes`, right?

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -80,72 +79,55 @@ public class MetadataLogConfig {
-        this.maxBatchSizeInBytes = KafkaRaftClient.MAX_BATCH_SIZE_BYTES;
-        this.maxFetchSizeInBytes = KafkaRaftClient.MAX_FETCH_SIZE_BYTES;
-        this.deleteDelayMillis = ServerLogConfigs.LOG_DELETE_DELAY_MS_DEFAULT;
+        this.maxBatchSizeInBytes = Objects.requireNonNullElse(config.getInt(INTERNAL_MAX_BATCH_SIZE_IN_BYTES_CONFIG), KafkaRaftClient.MAX_BATCH_SIZE_BYTES);
```
Should we name `maxBatchSizeInBytes` and the other two with the internal prefix? Also, should we just use KafkaRaftClient.MAX_BATCH_SIZE_BYTES as the default in the config def?

junrao commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -80,72 +79,55 @@ public class MetadataLogConfig {
 
     public int logSegmentBytes() {
-        return logSegmentBytes;
+        return Objects.requireNonNullElse(internalSegmentBytes, logSegmentBytes);
```
Hmm, it seems it's more intuitive to just return `logSegmentBytes` since we have a separate public api to return `internalSegmentBytes`.

chia7712 replied:
I think there are two styles:

```java
    public int logSegmentBytes() {
        return Objects.requireNonNullElse(internalSegmentBytes, logSegmentBytes);
    }
    
    public boolean useInternalSegmentBytes() {
        return internalSegmentBytes != null;
    }
    if (config.useInternalSegmentBytes())
      props.setProperty(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
    else
      props.setProperty(TopicConfig.SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
```


```java
    public int logSegmentBytes() {
        return logSegmentBytes;
    }
    
    public Integer internalSegmentBytes() {
        return internalSegmentBytes;
    }
    if (config.internalSegmentBytes() != null)
      props.setProperty(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, config.internalSegmentBytes.toString)
    else
      props.setProperty(TopicConfig.SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
```

I prefer  the later since `useInternalSegmentBytes` is not usual in config classes

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. A few more comments.

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

# Conflicts:
#	core/src/test/scala/unit/kafka/tools/DumpLogSegmentsTest.scala

m1a2st committed:
addressed by comments

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

junrao reviewed (commented):
@m1a2st : Thanks for the updated PR. The code LGTM. Are the test failures related?

m1a2st commented:
No, I think the fail tes are not related, but the main reason is the HashSet order is different, I will file a PR to fix it.

```
org.opentest4j.AssertionFailedError: expected: <[CoordinatorRecord(key=ShareGroupStatePartitionMetadataKey(groupId='groupId'), value=ApiMessageAndVersion(ShareGroupStatePartitionMetadataValue(initializingTopics=[], initializedTopics=[TopicPartitionsInfo(topicId=wO0yz3DfTaCC9fnfn8maXw, topicName='t1', partitions=[1, 0])], deletingTopics=[]) at version 0))]> but was: <[CoordinatorRecord(key=ShareGroupStatePartitionMetadataKey(groupId='groupId'), value=ApiMessageAndVersion(ShareGroupStatePartitionMetadataValue(initializingTopics=[], initializedTopics=[TopicPartitionsInfo(topicId=wO0yz3DfTaCC9fnfn8maXw, topicName='t1', partitions=[0, 1])], deletingTopics=[]) at version 0))]>
    at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
    at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
    at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
    at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
    at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
    at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1145)
    at app//org.apache.kafka.coordinator.group.GroupMetadataManagerTest.testShareGroupInitializeSuccess(GroupMetadataManagerTest.java:21717)
    at java.base@21.0.4/java.lang.reflect.Method.invoke(Method.java:580)
    at java.base@21.0.4/java.util.ArrayList.forEach(ArrayList.java:1596)
    at java.base@21.0.4/java.util.ArrayList.forEach(ArrayList.java:1596)
```

Update: https://issues.apache.org/jira/browse/KAFKA-19326
It already fix by https://github.com/apache/kafka/commit/af4d048da6223791c3e0ec51c739e9d324f60f82

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

m1a2st commented:
I merge trunk, and retrigger the CI again.

chia7712 commented on core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala:
```diff
@@ -1166,4 +1161,25 @@ object KafkaMetadataLogTest {
+    internalMaxFetchSizeInBytes: Int,
+    internalDeleteDelayMillis: Long
+  ): MetadataLogConfig = {
+    val config: util.Map[String, Any] = util.Map.of(
```
`val config = util.Map.of(`

m1a2st replied:
We need to explicitly declare entries as `Any`, otherwise the compiler will infer it as `long`, which would cause errors when parsing the config.

chia7712 commented on core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala:
```diff
@@ -1166,4 +1161,25 @@ object KafkaMetadataLogTest {
+    logSegmentMillis: Long,
+    retentionMaxBytes: Long,
+    retentionMillis: Long,
+    internalMaxBatchSizeInBytes: Int,
```
Could you please add default value to `internalMaxBatchSizeInBytes`, `internalMaxFetchSizeInBytes`, and `internalDeleteDelayMillis` to streamline the callers?

chia7712 commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -446,6 +447,14 @@ public static List<String> configNames() {
+                .sorted().toList();
+    }
+
     public static Optional<String> serverConfigName(String configName) {
```
it is unused now. please remove it

chia7712 commented on metadata/src/main/java/org/apache/kafka/metadata/KafkaConfigSchema.java:
```diff
@@ -166,6 +166,10 @@ public Map<String, ConfigEntry> resolveEffectiveTopicConfigs(
         HashMap<String, ConfigEntry> effectiveConfigs = new HashMap<>();
         for (ConfigDef.ConfigKey configKey : configDef.configKeys().values()) {
+            // This config is internal; if the user hasn't set it explicitly, it should not be returned.
+            if (configKey.internalConfig && !dynamicTopicConfigs.containsKey(configKey.name)) {
```
Have you added test for this change?

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LocalLogTest.java:
```diff
@@ -198,7 +197,7 @@ public void testUpdateConfig() {
 
         Properties props = new Properties();
-        props.put(TopicConfig.SEGMENT_BYTES_CONFIG, oldConfig.segmentSize + 1);
+        props.put(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, oldConfig.segmentSize() + 1);
```
it can keep using `TopicConfig.SEGMENT_BYTES_CONFIG`, right?

chia7712 commented on core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala:
```diff
@@ -95,29 +95,29 @@ class DynamicConfigChangeTest extends KafkaServerTestHarness {
     try {
       val resource = new ConfigResource(ConfigResource.Type.TOPIC, tp.topic())
-      val op = new AlterConfigOp(new ConfigEntry(TopicConfig.SEGMENT_BYTES_CONFIG, newSegmentSize.toString),
+      val op = new AlterConfigOp(new ConfigEntry(LogConfig.INTERNAL_SEGMENT_BYTES_CONFIG, newSegmentSize.toString),
```
this test is used to verify the alter, so you can use larger `newSegmentSize` to keep using `TopicConfig.SEGMENT_BYTES_CONFIG`

chia7712 reviewed (commented):
@m1a2st thanks for this patch. a couple of comments are left

m1a2st committed:
addressed by comments

m1a2st committed:
Merge branch 'trunk' into KAFKA-19080

chia7712 commented:
@m1a2st could you please tweak your PR description to summarize the changes?

m1a2st commented:
Thanks for reminder, updated it.

chia7712 commented:
@junrao @m1a2st Thanks for all your patient reviews and updates. I'm going to merge this. Please feel free to leave any further comments for follow-up.

mumrah commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -80,72 +78,55 @@ public class MetadataLogConfig {
             "controller should write no-op records to the metadata partition. If the value is 0, no-op records " +
             "are not appended to the metadata partition. The default value is " + METADATA_MAX_IDLE_INTERVAL_MS_DEFAULT;
 
+    public static final String INTERNAL_MAX_BATCH_SIZE_IN_BYTES_CONFIG = "internal.max.batch.size.in.bytes";
+    public static final String INTERNAL_MAX_BATCH_SIZE_IN_BYTES_DOC = "The largest record batch size allowed in the metadata log, only for testing.";
+
+    public static final String INTERNAL_MAX_FETCH_SIZE_IN_BYTES_CONFIG = "internal.max.fetch.size.in.bytes";
+    public static final String INTERNAL_MAX_FETCH_SIZE_IN_BYTES_DOC = "The maximum number of bytes to read when fetching from the metadata log, only for testing.";
+
+    public static final String INTERNAL_DELETE_DELAY_MILLIS_CONFIG = "internal.delete.delay.millis";
+    public static final String INTERNAL_DELETE_DELAY_MILLIS_DOC = "The amount of time to wait before deleting a file from the filesystem, only for testing.";
+
```
Shouldn't these configs have "metadata" in there somewhere in the name? From what I can tell, these are only used by KafkaMetadataLog.

mumrah commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -599,11 +601,7 @@ object KafkaMetadataLog extends Logging {
     LogConfig.validate(props)
     val defaultLogConfig = new LogConfig(props)
 
-    if (config.logSegmentBytes < config.logSegmentMinBytes) {
```
IIUC, this logic has been replaced by the config validation system.

SEGMENT_BYTES_CONFIG has minimum of 1 MB but can be overridden by INTERNAL_SEGMENT_BYTES_CONFIG which has no minimum.

mumrah commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -639,12 +637,6 @@ object KafkaMetadataLog extends Logging {
     )
 
-    // Print a warning if users have overridden the internal config
-    if (config.logSegmentMinBytes != KafkaRaftClient.MAX_BATCH_SIZE_BYTES) {
```
Should we keep the error message if INTERNAL_SEGMENT_BYTES_CONFIG has been set? This was added to help operators avoid setting these testing-only configs in a production environment.

mumrah commented on raft/src/main/java/org/apache/kafka/raft/MetadataLogConfig.java:
```diff
@@ -80,72 +78,55 @@ public class MetadataLogConfig {
-    public int logSegmentMinBytes() {
-        return logSegmentMinBytes;
+    
+    public Integer internalSegmentBytes() {
```
`OptionalInt` might be nicer here.

m1a2st replied:
Since `config.getInt(INTERNAL_METADATA_LOG_SEGMENT_BYTES_CONFIG)` may return a null value, using `OptionalInt` is not suitable here.
If we want to wrap the result with `Optional`, I think `Optional<Integer>` would be more appropriate.

mumrah reviewed (commented):
@m1a2st / @chia7712 -- some follow-up comments

chia7712 commented:
@mumrah thanks for your reviews. We will file a minor later

m1a2st commented:
I’ve addressed the follow-up comments on https://github.com/apache/kafka/pull/19842

chia7712 commented on core/src/main/scala/kafka/raft/KafkaMetadataLog.scala:
```diff
@@ -588,9 +588,11 @@ object KafkaMetadataLog extends Logging {
     val props = new Properties()
-    props.setProperty(TopicConfig.MAX_MESSAGE_BYTES_CONFIG, config.maxBatchSizeInBytes.toString)
-    props.setProperty(TopicConfig.SEGMENT_BYTES_CONFIG, config.logSegmentBytes.toString)
-    props.setProperty(TopicConfig.SEGMENT_MS_CONFIG, config.logSegmentMillis.toString)
```
this is a critical bug - open https://issues.apache.org/jira/browse/KAFKA-19392 to fix it
